docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,179 @@
+# Kubernetes compute resource metrics API
+
+## Goals
+
+Provide resource usage metrics on pods and nodes on the API server to be used
+by the scheduler to improve job placement, utilization, etc. and by end users
+to understand the resource utilization of their jobs.
+
+## Current state
+
+Right now, the Kubelet exports container metrics via an API endpoint. This
+information is not gathered nor served by the Kubernetes API server.
+
+## Use cases
+
+The first user will be kubectl. The resource usage data can be shown to the
+user via a periodically refreshing interface similar to `top` on Unix-like
+systems. This info could let users assign resource limits more efficiently.
+
+```
+$ kubectl top kubernetes-minion-abcd
+Pod: monitoring-heapster-abcde
+CPU:
+    mean: 0.12 cores
+    95th: 1.10 cores
+    max:  1.91 cores
+Memory:
+    mean: 300MB (10%)
+    95th: 330MB (11%)
+    max:  395MB (13%)
+```
+
+A second user will be the scheduler. To assign pods to nodes efficiently, the
+scheduler needs to know the current free resources on each node.
+
+## Proposed endpoints
+
+	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
```

Comment:
API group is orthogonal to subresource - you can do:

     /resourceapi/v1/nodes/foo/metrics

Just as easily.  Any push mechanism that's likely to be high volume is
something we might desire a separate component implement - having a
distinct group facilitates separation.  That said, I'd be just as happy
with

    /resourceapi/v1/rawnodemetrics/nodename



On Jul 28, 2015, at 5:21 PM, Derek Carr <notifications@github.com> wrote:

In docs/proposals/compute-resource-metrics-api.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11951#discussion_r35702782>
:

> +CPU:
> +    mean: 0.12 cores
> +    95th: 1.10 cores
> +    max:  1.91 cores
> +Memory:
> +    mean: 300MB (10%)
> +    95th: 330MB (11%)
> +    max:  395MB (13%)
> +```
> +
> +A second user will be the scheduler. To assign pods to nodes efficiently, the
> +scheduler needs to know the current free resources on each node.
> +
> +## Proposed endpoints
> +
> +	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]

I guess that was the part of the discussion that should go in a pros/cons
listing? It was not obvious to me that this should not be modeled as a
subresource, or why modeling as a subresource does what you imply.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11951/files#r35702782>
.